File: legacy/python/src/roostery/shim.py
"""lark-cli shim：流式 tee + TTY 直通 + journal envelope。

设计：``docs/FEISHU_OFFICE_HUB_DESIGN_V2.md`` §8。

关键约束：
- stdin/stdout/stderr 任一为 TTY，或 argv[0] 命中 ``interactive_verbs`` → ``os.execv`` 直通；
- 非交互：``Popen`` + 两个 pump 线程实时 tee；
- shim 与 real 路径相同 → 立即报错，防递归调用。
"""
from __future__ import annotations

import os
import subprocess
import sys
import threading
import time
from pathlib import Path
from typing import Any, Callable, Dict, Iterable, List, Optional, Sequence, Tuple

from . import config as cfgmod
from . import journal, redact, remoterefs

NO_JOURNAL_ENV = "FEISHU_HUB_NOJOURNAL"


def is_interactive(
    argv: Sequence[str],
    interactive_verbs: Iterable[str],
    *,
    isatty_fn: Optional[Callable[[int], bool]] = None,
) -> bool:
    """判断是否走 ``os.execv`` 直通路径。"""
    check = isatty_fn if isatty_fn is not None else os.isatty
    try:
        if any(check(fd) for fd in (0, 1, 2)):
            return True
    except OSError:
        pass
    if argv and argv[0] in set(interactive_verbs):
        return True
    if any(a in ("--interactive", "--repl", "-i") for a in argv):
        return True
    return False


def resolve_real_cli(cfg: Dict[str, Any], *, shim_path: str) -> str:
    """从 config + env 取真实 lark-cli 路径，做防递归校验。"""
    real = cfg.get("shim", {}).get("real_lark_cli", "") or ""
    if not real:
        raise RuntimeError(
            "roostery: shim.real_lark_cli not configured; "
            "run `python -m roostery init` or set FEISHU_HUB_REAL_LARK_CLI"
        )
    real_resolved = os.path.realpath(real)
    shim_resolved = os.path.realpath(shim_path)
    if real_resolved == shim_resolved:
        raise RuntimeError(
            f"roostery: real_lark_cli ({real}) resolves to shim itself; abort"
        )
    if not os.path.exists(real_resolved):
        raise RuntimeError(f"roostery: real_lark_cli not found: {real}")
    return real_resolved


class _Null:
    def write(self, _): return None
    def flush(self): return None


def _bin_stream(s):
    """取得二进制可写流；pytest capture 等没有 .buffer 时退化为 sink。"""
    buf = getattr(s, "buffer", None)
    if buf is not None:
        return buf
    return _Null()


def _pump(src, dst_bin, cap: int, sink: bytearray) -> None:
    """从 src 读出，同时写入 dst_bin 与 head sink。"""
    while True:
        chunk = src.read(4096)
        if not chunk:
            break
        try:
            dst_bin.write(chunk)
            dst_bin.flush()
        except (BrokenPipeError, ValueError):
            pass
        if len(sink) < cap:
            need = cap - len(sink)
            sink.extend(chunk[:need])


def run_non_interactive(
    real_cli: str,
    sub_argv: Sequence[str],
    *,
    stdout_head_cap: int,
    stderr_head_cap: int,
    stdin=None,
    stdout=None,
    stderr=None,
) -> Tuple[int, bytes, bytes, int]:
    """运行非交互命令，返回 (rc, stdout_head, stderr_head, duration_ms)。"""
    if stdin is None:
        stdin = sys.stdin
        # 被 pytest / 非 TTY 包裹时，sys.stdin 可能没有 fileno；
        # 这种情形对 lark-cli 等价于"无 stdin"，直接给 DEVNULL。
        try:
            stdin.fileno()
        except (AttributeError, OSError, ValueError):
            stdin = subprocess.DEVNULL
    out_bin = _bin_stream(stdout if stdout is not None else sys.stdout)
    err_bin = _bin_stream(stderr if stderr is not None else sys.stderr)

    t0 = time.time()
    proc = subprocess.Popen(
        [real_cli, *list(sub_argv)],
        stdin=stdin,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        bufsize=0,
    )
    stdout_head = bytearray()
    stderr_head = bytearray()
    t_out = threading.Thread(
        target=_pump,
        args=(proc.stdout, out_bin, stdout_head_cap, stdout_head),
    )
    t_err = threading.Thread(
        target=_pump,
        args=(proc.stderr, err_bin, stderr_head_cap, stderr_head),
    )
    t_out.start()
    t_err.start()
    rc = proc.wait()
    t_out.join()
    t_err.join()
    duration_ms = int((time.time() - t0) * 1000)
    return rc, bytes(stdout_head), bytes(stderr_head), duration_ms


def build_record(
    sub_argv: Sequence[str],
    rc: int,
    stdout_head: bytes,
    stderr_head: bytes,
    duration_ms: int,
    *,
    stdin_present: bool,
) -> Dict[str, Any]:
    argv_red, redacted = redact.scrub_argv(sub_argv)
    stdout_red = redact.scrub_text(stdout_head)
    stderr_red = redact.scrub_text(stderr_head)
    return {
        "event_type": "lark_cli.invoke",
        "source": "shim",
        "actor": journal.actor_from_env(),
        "cwd": os.getcwd(),
        "command": {
            "argv": argv_red,
            "duration_ms": duration_ms,
            "exit_code": rc,
        },
        "io": {
            "stdout_head": stdout_red,
            "stderr_head": stderr_red,
            "stdin_present": stdin_present,
            "tty": False,
        },
        "remote_refs": remoterefs.extract(list(sub_argv), stdout_head),
        "summary": None,
        "tags": journal.tags_from_env(),
        "privacy": {"redacted_fields": redacted, "no_journal_reason": None},
    }


def main(argv: Optional[Sequence[str]] = None) -> int:
    """shim 入口；返回 real lark-cli 的退出码（异常时返回 127）。"""
    if argv is None:
        argv = sys.argv
    sub_argv = list(argv[1:])
    shim_path = argv[0]

    try:
        cfg = cfgmod.load()
        real_cli = resolve_real_cli(cfg, shim_path=shim_path)
    except RuntimeError as e:
        sys.stderr.write(f"[roostery] {e}\n")
        return 127

    no_journal = os.getenv(NO_JOURNAL_ENV) == "1"
    interactive_verbs = cfg["shim"].get("interactive_verbs", [])

    if is_interactive(sub_argv, interactive_verbs):
        if not no_journal:
            try:
                journal.append_skipped(sub_argv, reason="interactive")
            except Exception as e:
                sys.stderr.write(f"[roostery] journal skipped-write failed: {e}\n")
        os.execv(real_cli, [real_cli, *sub_argv])  # 不返回
        return 0  # 仅为类型完整

    stdin_present = False
    try:
        stdin_present = not sys.stdin.isatty()
    except (AttributeError, ValueError):
        pass

    try:
        rc, out_head, err_head, dur = run_non_interactive(
            real_cli,
            sub_argv,
            stdout_head_cap=int(cfg["shim"]["stdout_head_bytes"]),
            stderr_head_cap=int(cfg["shim"]["stderr_head_bytes"]),
        )
    except FileNotFoundError:
        sys.stderr.write(f"[roostery] real lark-cli not found: {real_cli}\n")
        return 127

    if no_journal:
        try:
            journal.append_skipped(sub_argv, reason="env")
        except Exception as e:
            sys.stderr.write(f"[roostery] journal skipped-write failed: {e}\n")
    else:
        try:
            rec = build_record(sub_argv, rc, out_head, err_head, dur,
                               stdin_present=stdin_present)
            journal.append(rec)
        except Exception as e:
            sys.stderr.write(f"[roostery] journal write failed: {e}\n")

    return rc


if __name__ == "__main__":  # pragma: no cover
    raise SystemExit(main(sys.argv))

File: crates/roostery/src/bin/shim.rs
//! `lark-cli` PATH-prefix shim — intercepts agent-runtime calls to `lark-cli`,
//! tees stdout/stderr to the user while writing a [`JournalEntry`] for audit,
//! then delegates to the real `lark-cli` binary.
//!
//! See `.codestable/features/2026-05-17-lark-cli-shim/lark-cli-shim-design.md`
//! for the full behavior contract.

use roostery::journal::{JournalEntry, JournalResult};
use roostery::redact;
use roostery::remoterefs;
use serde_json::json;
use std::io::IsTerminal;
use std::path::{Path, PathBuf};
use std::process::ExitCode;
use thiserror::Error;

const ENV_REAL_CLI: &str = "ROOSTERY_REAL_LARK_CLI";
const ENV_NOJOURNAL: &str = "ROOSTERY_NOJOURNAL";
const INTERACTIVE_VERBS: &[&str] = &["auth"];
const STDOUT_HEAD_CAP: usize = 64 * 1024;
const STDERR_HEAD_CAP: usize = 16 * 1024;

#[derive(Debug, Error)]
enum ShimError {
    #[error("ROOSTERY_REAL_LARK_CLI not set; run `roostery init`")]
    MissingRealCli,
    #[error(
        "real_lark_cli ({real:?}) resolves to shim itself ({shim:?}); abort to prevent recursion"
    )]
    Recursion { real: PathBuf, shim: PathBuf },
    #[error("real_lark_cli not found: {path:?}")]
    RealCliNotFound { path: PathBuf },
    #[error("journal append failed: {source}")]
    #[allow(dead_code)]
    JournalFailed { source: std::io::Error },
}

/// Resolve the real `lark-cli` path from env and guard against the shim
/// pointing at itself (would cause infinite recursion).
fn resolve_real_cli() -> Result<PathBuf, ShimError> {
    let raw = std::env::var_os(ENV_REAL_CLI).filter(|v| !v.is_empty());
    let raw = raw.ok_or(ShimError::MissingRealCli)?;
    let real = PathBuf::from(raw);
    if !real.exists() {
        return Err(ShimError::RealCliNotFound { path: real });
    }
    let real_canon = std::fs::canonicalize(&real)
        .map_err(|_| ShimError::RealCliNotFound { path: real.clone() })?;
    if let Ok(self_exe) = std::env::current_exe()
        && let Ok(self_canon) = std::fs::canonicalize(&self_exe)
        && real_canon == self_canon
    {
        return Err(ShimError::Recursion {
            real: real_canon,
            shim: self_canon,
        });
    }
    Ok(real_canon)
}

/// Decide whether the call must go through `execv` (interactive direct-through).
/// Three-stage check: TTY on any std fd, verb in [`INTERACTIVE_VERBS`], or
/// presence of `--interactive` / `-i` / `--repl` flag.
fn is_interactive(sub_argv: &[String]) -> bool {
    if std::io::stdin().is_terminal()
        || std::io::stdout().is_terminal()
        || std::io::stderr().is_terminal()
    {
        return true;
    }
    if let Some(verb) = sub_argv.first()
        && INTERACTIVE_VERBS.contains(&verb.as_str())
    {
        return true;
    }
    sub_argv
        .iter()
        .any(|a| a == "--interactive" || a == "-i" || a == "--repl")
}

/// Variant of `is_interactive` that ignores TTY state — testable in any env.
#[cfg(test)]
fn is_interactive_argv(sub_argv: &[String]) -> bool {
    if let Some(verb) = sub_argv.first()
        && INTERACTIVE_VERBS.contains(&verb.as_str())
    {
        return true;
    }
    sub_argv
        .iter()
        .any(|a| a == "--interactive" || a == "-i" || a == "--repl")
}

enum Outcome {
    Full {
        rc: i32,
        stdout_head: Vec<u8>,
        stderr_head: Vec<u8>,
        duration_ms: u64,
        stdin_present: bool,
    },
    Skipped {
        reason: &'static str,
    },
}

/// Pump bytes from `src` to `dst` while keeping the first `cap` bytes in
/// `head`. Writes to `dst` are best-effort (broken pipes are swallowed) so
/// transparency wins over journal completeness.
fn pump<R: std::io::Read, W: std::io::Write>(
    mut src: R,
    mut dst: W,
    cap: usize,
) -> std::io::Result<Vec<u8>> {
    let mut head = Vec::with_capacity(cap.min(4096));
    let mut buf = [0u8; 4096];
    loop {
        let n = src.read(&mut buf)?;
        if n == 0 {
            break;
        }
        let chunk = &buf[..n];
        let _ = dst.write_all(chunk);
        let _ = dst.flush();
        if head.len() < cap {
            let take = cap - head.len();
            head.extend_from_slice(&chunk[..n.min(take)]);
        }
    }
    Ok(head)
}

/// Run the real `lark-cli` non-interactively: spawn, two pump threads
/// (stdout / stderr), wait, return (rc, stdout_head, stderr_head, duration_ms).
fn run_non_interactive(
    real_cli: &Path,
    sub_argv: &[String],
) -> std::io::Result<(i32, Vec<u8>, Vec<u8>, u64)> {
    use std::process::{Command, Stdio};
    use std::time::Instant;

    let started = Instant::now();
    let mut child = Command::new(real_cli)
        .args(sub_argv)
        .stdin(Stdio::inherit())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()?;

    let child_stdout = child.stdout.take().expect("stdout piped");
    let child_stderr = child.stderr.take().expect("stderr piped");

    let t_out = std::thread::spawn(move || pump(child_stdout, std::io::stdout(), STDOUT_HEAD_CAP));
    let t_err = std::thread::spawn(move || pump(child_stderr, std::io::stderr(), STDERR_HEAD_CAP));

    let status = child.wait()?;
    let stdout_head = t_out.join().unwrap_or(Ok(Vec::new())).unwrap_or_default();
    let stderr_head = t_err.join().unwrap_or(Ok(Vec::new())).unwrap_or_default();
    let rc = status.code().unwrap_or(1);
    let duration_ms = started.elapsed().as_millis() as u64;
    Ok((rc, stdout_head, stderr_head, duration_ms))
}

/// Build a `JournalEntry` for either a Full run or a Skipped placeholder.
/// Maps onto the 11-field schema (see roadmap §4.2); extras live under `params`.
fn build_entry(sub_argv: &[String], outcome: Outcome) -> JournalEntry {
    let verb = sub_argv
        .first()
        .cloned()
        .unwrap_or_else(|| "<empty>".into());
    let (scrubbed_argv, _) = redact::scrub_argv(sub_argv);
    match outcome {
        Outcome::Full {
            rc,
            stdout_head,
            stderr_head,
            duration_ms,
            stdin_present,
        } => {
            let stdout_str = String::from_utf8_lossy(&stdout_head);
            let stderr_str = String::from_utf8_lossy(&stderr_head);
            let stdout_clean = redact::scrub_text(&stdout_str);
            let stderr_clean = redact::scrub_text(&stderr_str);
            let refs = remoterefs::extract(sub_argv, &stdout_str);
            let refs_value = serde_json::to_value(&refs).unwrap_or(serde_json::Value::Null);
            let cwd = std::env::current_dir()
                .ok()
                .map(|p| p.to_string_lossy().into_owned())
                .unwrap_or_default();
            let mut e = JournalEntry::new("shim", format!("lark-cli:{}", verb));
            e.params = json!({
                "argv": scrubbed_argv,
                "cwd": cwd,
                "stdin_present": stdin_present,
                "stdout_head": stdout_clean,
                "stderr_head": stderr_clean,
                "remote_refs": refs_value.clone(),
            });
            e.duration_ms = duration_ms;
            e.result = if rc == 0 {
                JournalResult::Ok { value: refs_value }
            } else {
                JournalResult::Err {
                    kind: "NonZeroExit".into(),
                    message: format!(
                        "exit={} stderr={}",
                        rc,
                        stderr_clean.lines().next().unwrap_or("")
                    ),
                }
            };
            e
        }
        Outcome::Skipped { reason } => {
            let mut e = JournalEntry::new("shim", format!("lark-cli:{}:skipped", verb));
            e.params = json!({
                "argv": scrubbed_argv,
                "reason": reason,
            });
            e.duration_ms = 0;
            e
        }
    }
}

fn main() -> ExitCode {
    let sub_argv: Vec<String> = std::env::args().skip(1).collect();

    let real_cli = match resolve_real_cli() {
        Ok(p) => p,
        Err(e) => {
            eprintln!("[roostery] {e}");
            return ExitCode::from(127);
        }
    };

    let journal = roostery::journal::Journal::default();

    if is_interactive(&sub_argv) {
        let entry = build_entry(
            &sub_argv,
            Outcome::Skipped {
                reason: "interactive",
            },
        );
        if let Err(e) = journal.append(&entry) {
            tracing::warn!(error = %e, "shim: journal append failed (interactive)");
        }
        use std::os::unix::process::CommandExt as _;
        let err = std::process::Command::new(&real_cli).args(&sub_argv).exec();
        // exec only returns on failure.
        eprintln!("[roostery] exec failed: {err}");
        return ExitCode::from(127);
    }

    let nojournal = matches!(std::env::var(ENV_NOJOURNAL).as_deref(), Ok("1"));

    let (rc, stdout_head, stderr_head, duration_ms) =
        match run_non_interactive(&real_cli, &sub_argv) {
            Ok(t) => t,
            Err(e) => {
                eprintln!("[roostery] spawn failed: {e}");
                return ExitCode::from(127);
            }
        };

    let entry = if nojournal {
        build_entry(
            &sub_argv,
            Outcome::Skipped {
                reason: "nojournal",
            },
        )
    } else {
        let stdin_present = !std::io::stdin().is_terminal();
        build_entry(
            &sub_argv,
            Outcome::Full {
                rc,
                stdout_head,
                stderr_head,
                duration_ms,
                stdin_present,
            },
        )
    };

    if let Err(e) = journal.append(&entry) {
        tracing::warn!(error = %e, "shim: journal append failed");
    }

    rc_to_exitcode(rc)
}

fn rc_to_exitcode(rc: i32) -> ExitCode {
    if let Ok(byte) = u8::try_from(rc) {
        ExitCode::from(byte)
    } else {
        ExitCode::from(1)
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    // codex audit finding-09: 统一用 crate-wide TEST_ENV_LOCK。shim 是 cargo bin
    // target，同 crate 的 lib 模块可直接 `use roostery::...` 访问。
    use roostery::paths::TEST_ENV_LOCK as ENV_LOCK;

    fn with_env<F: FnOnce()>(real: Option<&str>, body: F) {
        let _g = ENV_LOCK.lock().unwrap();
        // Safety: tests touching env are serialized via ENV_LOCK.
        unsafe {
            match real {
                Some(v) => std::env::set_var(ENV_REAL_CLI, v),
                None => std::env::remove_var(ENV_REAL_CLI),
            }
        }
        body();
        unsafe { std::env::remove_var(ENV_REAL_CLI) };
    }

    #[test]
    fn resolve_missing_env() {
        with_env(None, || {
            assert!(matches!(resolve_real_cli(), Err(ShimError::MissingRealCli)));
        });
    }

    #[test]
    fn resolve_empty_env_is_missing() {
        with_env(Some(""), || {
            assert!(matches!(resolve_real_cli(), Err(ShimError::MissingRealCli)));
        });
    }

    #[test]
    fn resolve_nonexistent_path() {
        with_env(Some("/definitely/not/here/lark-cli"), || {
            assert!(matches!(
                resolve_real_cli(),
                Err(ShimError::RealCliNotFound { .. })
            ));
        });
    }

    #[test]
    fn resolve_happy_path() {
        // /bin/sh exists on every Unix; canonicalize works.
        with_env(Some("/bin/sh"), || {
            let got = resolve_real_cli().expect("happy");
            assert!(got.exists());
            assert!(got.is_absolute());
        });
    }

    #[test]
    fn resolve_recursion_detected() {
        // Point ENV_REAL_CLI at current_exe(); canonicalize equal → Recursion.
        let self_exe = std::env::current_exe().expect("test binary");
        with_env(Some(self_exe.to_str().unwrap()), || {
            assert!(matches!(
                resolve_real_cli(),
                Err(ShimError::Recursion { .. })
            ));
        });
    }

    #[test]
    fn is_interactive_truth_table() {
        let s = |v: &[&str]| v.iter().map(|x| x.to_string()).collect::<Vec<_>>();
        assert!(is_interactive_argv(&s(&["auth", "login"])));
        assert!(is_interactive_argv(&s(&["any", "--interactive"])));
        assert!(is_interactive_argv(&s(&["any", "-i"])));
        assert!(is_interactive_argv(&s(&["any", "--repl"])));
        assert!(!is_interactive_argv(&s(&["im", "+messages-send"])));
    }

    #[test]
    fn build_entry_full_schema_locked() {
        let argv = vec!["im".into(), "+messages-send".into()];
        let entry = build_entry(
            &argv,
            Outcome::Full {
                rc: 0,
                stdout_head: br#"{"data":{"message_id":"om_abc","chat_id":"oc_x"}}"#.to_vec(),
                stderr_head: Vec::new(),
                duration_ms: 12,
                stdin_present: false,
            },
        );
        assert_eq!(entry.schema_version, roostery::SCHEMA_VERSION);
        assert_eq!(entry.source, "shim");
        assert_eq!(entry.action, "lark-cli:im");
        assert_eq!(entry.duration_ms, 12);
        let p = &entry.params;
        for key in [
            "argv",
            "cwd",
            "stdin_present",
            "stdout_head",
            "stderr_head",
            "remote_refs",
        ] {
            assert!(p.get(key).is_some(), "params missing {key}");
        }
        match entry.result {
            JournalResult::Ok { value } => {
                assert_eq!(value["message_id"], "om_abc");
            }
            JournalResult::Err { .. } => panic!("expected Ok"),
        }
    }

    #[test]
    fn build_entry_full_nonzero_is_err() {
        let entry = build_entry(
            &["docs".into()],
            Outcome::Full {
                rc: 7,
                stdout_head: Vec::new(),
                stderr_head: b"boom\n".to_vec(),
                duration_ms: 1,
                stdin_present: false,
            },
        );
        match entry.result {
            JournalResult::Err { kind, message } => {
                assert_eq!(kind, "NonZeroExit");
                assert!(message.contains("exit=7"));
            }
            JournalResult::Ok { .. } => panic!("expected Err"),
        }
    }

    #[test]
    fn build_entry_skipped_schema() {
        let entry = build_entry(
            &["auth".into(), "login".into()],
            Outcome::Skipped {
                reason: "interactive",
            },
        );
        assert_eq!(entry.source, "shim");
        assert_eq!(entry.action, "lark-cli:auth:skipped");
        assert_eq!(entry.duration_ms, 0);
        assert_eq!(entry.params["reason"], "interactive");
        assert!(matches!(entry.result, JournalResult::Ok { .. }));
    }

    #[test]
    fn build_entry_edge_cases() {
        let e1 = build_entry(
            &[],
            Outcome::Skipped {
                reason: "nojournal",
            },
        );
        assert_eq!(e1.action, "lark-cli:<empty>:skipped");

        let argv = vec![
            "im".into(),
            "--access-token".into(),
            "xyz".into(),
            "send".into(),
        ];
        let e2 = build_entry(
            &argv,
            Outcome::Full {
                rc: 0,
                stdout_head: Vec::new(),
                stderr_head: Vec::new(),
                duration_ms: 1,
                stdin_present: false,
            },
        );
        assert_eq!(e2.params["argv"].as_array().unwrap()[2], "***");
    }

    fn fixture_script(body: &str) -> (tempfile::TempDir, PathBuf) {
        use std::os::unix::fs::PermissionsExt as _;
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("fake-lark-cli");
        let mut content = String::from("#!/bin/sh\n");
        content.push_str(body);
        std::fs::write(&path, content).unwrap();
        let mut perm = std::fs::metadata(&path).unwrap().permissions();
        perm.set_mode(0o755);
        std::fs::set_permissions(&path, perm).unwrap();
        (dir, path)
    }

    #[test]
    fn run_happy_and_exit_passthrough() {
        let (_d, p) = fixture_script("echo hello && echo err >&2 && exit 0");
        let (rc, out, err, _) = run_non_interactive(&p, &[]).unwrap();
        assert_eq!(
            (rc, &out[..], &err[..]),
            (0, &b"hello\n"[..], &b"err\n"[..])
        );
        let (_d2, p2) = fixture_script("exit 42");
        assert_eq!(run_non_interactive(&p2, &[]).unwrap().0, 42);
    }

    #[test]
    fn run_head_caps() {
        let (_d, p) = fixture_script("head -c 204800 /dev/zero | tr '\\0' A");
        let (_, out, _, _) = run_non_interactive(&p, &[]).unwrap();
        assert_eq!(out.len(), STDOUT_HEAD_CAP);
        assert!(out.iter().all(|b| *b == b'A'));
        let (_d2, p2) = fixture_script("head -c 65536 /dev/zero | tr '\\0' B >&2");
        assert_eq!(
            run_non_interactive(&p2, &[]).unwrap().2.len(),
            STDERR_HEAD_CAP
        );
    }
}
